File: Tracks/typescript/complex-numbers/complex-numbers.ts
export default class ComplexNumber {
    real: number = 0;
    imag: number = 0;

    constructor(real: number, imag: number) {
        this.real = real;
        this.imag = imag;
    }

    add(other: ComplexNumber): ComplexNumber {
        return new ComplexNumber(this.real + other.real, this.imag + other.imag);
    }
    sub(other: ComplexNumber): ComplexNumber {
        return new ComplexNumber(this.real - other.real, this.imag - other.imag);
    }
    mul(other: ComplexNumber): ComplexNumber {
        return new ComplexNumber(
            this.real * other.real - this.imag * other.imag,
            this.imag * other.real + this.real * other.imag
        );
    }
    div(other: ComplexNumber): ComplexNumber {
        const denominator = other.real ** 2 + other.imag ** 2;
        return new ComplexNumber(
            (this.real * other.real + this.imag * other.imag) / denominator,
            (this.imag * other.real - this.real * other.imag) / denominator
        );
    }
    abs(): number {
        return Math.sqrt(this.real ** 2 + this.imag ** 2);
    }
    conj(): ComplexNumber {
        return new ComplexNumber(this.real, this.imag === 0 ? 0 : -1 * this.imag);
    }
    exp(): ComplexNumber {
        const expReal = Math.exp(this.real);
        return new ComplexNumber(expReal * Math.cos(this.imag), expReal * Math.sin(this.imag));
    }
}

File: Tracks/csharp/complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private readonly double real;
    private readonly double imaginary;

    public ComplexNumber(double real, double imaginary)
    {
        this.real = real;
        this.imaginary = imaginary;
    }

    public double Real() => real;

    public double Imaginary() => imaginary;

    public ComplexNumber Mul(ComplexNumber other) => new ComplexNumber(real * other.real - imaginary * other.imaginary, imaginary * other.real + real * other.imaginary);

    public ComplexNumber Add(ComplexNumber other) => new ComplexNumber(real + other.real, imaginary + other.imaginary);

    public ComplexNumber Sub(ComplexNumber other) => new ComplexNumber(real - other.real, imaginary - other.imaginary);

    public ComplexNumber Div(ComplexNumber other)
    {
        double denominator = other.imaginary * other.imaginary + other.real * other.real;
        return Mul(other.Conjugate()) / denominator;
    }

    public static ComplexNumber operator /(ComplexNumber complex, double number) => new ComplexNumber(complex.real / number, complex.imaginary / number);
    public static ComplexNumber operator *(ComplexNumber complex, double number) => new ComplexNumber(complex.real * number, complex.imaginary * number);

    public double Abs() => Math.Sqrt(real * real + imaginary * imaginary);

    public ComplexNumber Conjugate() => new ComplexNumber(real, -1 * imaginary);

    public ComplexNumber Exp()
    {
        double commonFactor = Math.Exp(real);
        return new ComplexNumber(Math.Cos(imaginary), Math.Sin(imaginary)) * commonFactor;
    }
}